### PR TITLE
dialects: add pdl custom format

### PR DESCRIPTION
**Pull request by math-fehr**
This PR adds custom format for PDL operations.
It also fix 1-2 bugs in the parser.
The format is tested in the updated PDL tests, and a conversion from custom and generic format is also tested in a new test.

### CODE REVIEW

**Review by tobiasgrosser — APPROVED** [review]
Very nice!

File: xdsl/dialects/pdl.py [hunk 1]
```diff
@@ -34,10 +34,44 @@
...
+        Parser.Delimiter.NONE, parser.parse_attribute
+    )
+    end_pos = parser.pos
+    if len(operands) != len(types):
+        parser.raise_error(
+            "Mismatched between the numbers of operands and types", pos, end_pos
+        )
+    for operand, type in zip(operands, types):
+        if operand.typ != type:
+            raise Exception("Mismatched between operands and their types", pos, end_pos)
```
Comment (superlopuh) [inline xdsl/dialects/pdl.py:64]: Seems like this should be a subtype, why not `parser.raise_error` here also?

**Review by superlopuh — APPROVED** [review]

File: tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir
Comment (webmiche) [inline tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir:1]: Is there really actually no better name for this test? xD
(same for the next few)

File: tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
-// CHECK-NEXT:    "pdl.pattern"() ({
-// CHECK-NEXT:      %0 = "pdl.type"() {"constantType" = i32} : () -> !pdl.type
-// CHECK-NEXT:      %1 = "pdl.type"() : () -> !pdl.type
-// CHECK-NEXT:      %2 = "pdl.operation"(%0, %1) {"attributeValueNames" = [], "operand_segment_sizes" = array<i32: 0, 0, 2>} : (!pdl.type, !pdl.type) -> !pdl.operation
-// CHECK-NEXT:      "pdl.rewrite"(%2) ({
-// CHECK-NEXT:        %3 = "pdl.operation"(%0, %1) {"attributeValueNames" = [], "opName" = "foo.op", "operand_segment_sizes" = array<i32: 0, 0, 2>} : (!pdl.type, !pdl.type) -> !pdl.operation
-// CHECK-NEXT:      }) {"operand_segment_sizes" = array<i32: 1, 0>} : (!pdl.operation) -> ()
-// CHECK-NEXT:    }) {"benefit" = 1 : i16, "sym_name" = "infer_type_from_type_used_in_match"} : () -> ()
-// CHECK-NEXT:  }) : () -> ()
+// CHECK:      builtin.module {
```
Comment (webmiche) [inline tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir:16]: In a lot of these you are checking all of the IR prints. Is this really necessary? I feel at least the module line is useless, right?

This leads to bigger conflict surfaces with changes in dependent parts of the project, which I feel is bad practice.

**Review by webmiche — CHANGES_REQUESTED** [review]
Can you split this PR? A trivial one on adding generic syntax and one for the pdl stuff?

File: tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir (hunk 2 above)
Comment (superlopuh) [reply tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir:16]: I think the changes got conflated with changes already merged, it just needs a rebase, the diff was much smaller previously

File: tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir
Comment (math-fehr) [reply tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir:1]: These tests were extracted from a large `split-input-file` file.
Now that we support that, I put them back together, so no need for names anymore ;)

File: tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir (hunk 2 above)
Comment (math-fehr) [reply tests/filecheck/dialects/pdl/mlir-tests/infer_type_from_type_used_in_match_0.mlir]: I removed the `builtin.module` check, yes

File: xdsl/dialects/pdl.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/dialects/pdl.py]: Thanks! I changed it to `raise_error`.

**Pull request merged**